README.md

# Logs - Laravel

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

## Install

Via Composer

``` bash
$ composer require dev4born/logs
```

## Integration

In the `$providers` array add the `service providers` for this package.

```
Dev4Born\logs\LogsLaravelServiceProvider::class
```

Create config file `config/logs.php` and add the following lines.

```
/**
 *   
 *  Warning! You can only grant one permissions from 2 available
 *  
 *  @package dev4born\logs
 *  
 */
 
return [

    'settings' => [
	
    /**
	 *  
	 *  Secret key - unique address for accessing logs
	 *  
	 *  http://{your-project}/laravel/{secret-key}/logs
	 *  
	 */
	
        'secret' => '',

    /**
	 *  
	 *  Allowed permissions (middleware) for accessing logs
	 *  
	 *  http://{your-project}/laravel/view/logs
	 *  
	 */   
		
        'middleware' => [
            
        ],
		
    ],
	
];
```

You can specify 2 permissions (secret key or middleware) for accessing logs.

## Usage

Secret key

http://{your-project}/laravel/{your-secret-key}/logs 

Middleware

http://{your-project}/laravel/view/logs 

## JSON API

Secret key

http://{your-project}/laravel/{your-secret-key}/logs/json

Middleware

http://{your-project}/laravel/view/logs/json

**Output**

There are no problems.

```
{'error':'Everything functions properly.'}
```

There was an error.

```
{'error':'Perhaps an error occurred - check the logs/events.'}
```

Messages can be implemented in the dashboard.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email: milosz.nowak@dev4born.pro instead of using the issue tracker.

## Credits

- [Dev4Born][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Screenshots

![Dashboard](https://dev4born.pro/pub/github/logs_screenshot.png)

[ico-version]: https://img.shields.io/packagist/v/dev4born/logs.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/dev4born/logs.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/dev4born/logs
[link-downloads]: https://packagist.org/packages/dev4born/logs
[link-author]: https://github.com/dev4born
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `logs` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## Date of start: 13.12.2017

### (1.0.5)
- Add new permission method
- Visual corrections
- Fixed routing

### (1.0.0)
- Visual corrections

src/Views/logs.blade.php
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <title>Logs - Laravel</title>

    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.2/css/bootstrap.min.css" rel="stylesheet">
	<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
	<link href="https://dev4born.pro/pub/css/bootstrap.css" rel="stylesheet">
	
    <style type="text/css">
        body {
            padding-top: 54px;
			color: #434343;
        }
	    .bg-color {
		    background-color: #434343!important;  
	    }
		.pre-logs {
			text-align: left; 
			height: 500px;	
		}
    </style>
	
	<script type="text/javascript">
	    function scroll_pre() {
            var scrolled = document.getElementById('pre');
            scrolled.scrollTop = scrolled.scrollHeight;
        }
    </script>
  </head>
  <body onload="scroll_pre();">
    <nav class="navbar-expand-lg navbar-dark bg-color fixed-top">
      <div class="container">
        <a class="navbar-brand" href="{{ url('laravel/'.$request->route('secret').'/logs') }}">Logs - Laravel</a>
      </div>
    </nav>
    <div class="container">
      <div class="row">
        <div class="col-lg-12 text-center">
            <h1 class="mt-5">List of files</h1>
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>Name</th>
						<th>Modified</th>
                        <th>Actions</th>
                    </tr>
                </thead>
                <tbody>
	        @foreach($list as $file)
                    <tr>
                        <td>{{ basename($file) }}</td>
						<td>{{ date('Y-m-d H:i:s', filemtime($file)) }}</td>
                        <td>
		                    <a href="{{ url('laravel/'.$request->route('secret').'/logs?file='.basename($file)) }}" class="btn btn-default btn-sm">
                                <i class="fa fa-eye"></i> Show
                            </a>			
		                    <a href="{{ url('laravel/logs/'.sha1(date('Y-m-d H:i')).'_'.basename($file)) }}" class="btn btn-default btn-sm">
                                <i class="fa fa-download"></i> Download
                            </a>
		                    <a href="{{ url('laravel/logs/remove/'.sha1(date('Y-m-d H:i')).'_'.basename($file)) }}" class="btn btn-default btn-sm">
                                <i class="fa fa-remove"></i> Remove
                            </a>		
		                </td>
                    </tr>
	        @endforeach
                </tbody>
            </table> 		
            <h1 class="mt-5">All events - @if($request['file']) {{ $request['file'] }} @else laravel.log @endif</h1>    
		@if($logs)
			<pre class="pre-logs" id="pre">	
			    {{ $logs }}
			</pre> 
		@else
			<div class="alert alert-warning">
                <strong>Info!</strong> Unfortunately, no logs/events were found.
            </div>
		@endif
            </pre> 
            <div class="list-unstyled">
                <p>Powered by <strong>Dev4Born.pro</strong></p>
            </div>
        </div>
      </div>
    </div>
  </body>
</html>


src/LogsLaravelServiceProvider.php
<?php

/**
 *  
 *  Copyright (C) 2017
 *  
 *  File      : LogsLaravelServiceProvider.php
 *  Author    : Miłosz Nowak
 *  Copyright : (c) 2017 Dev4Born
 *  Link      : https://dev4born.pro
 *  Date      : 12/13/17
 *  
 */
 
namespace dev4born\logs;

use Illuminate\Support\ServiceProvider;
use Illuminate\Routing\Router;

/**
 *  
 *  Provider LogsLaravelServiceProvider
 *  
 *  
 *  @package dev4born\logs
 *  
 */
 
class LogsLaravelServiceProvider extends ServiceProvider
{
    /**
     *  
     * @brief Indicates if loading of the provider is deferred
     *
	 *
     * @var bool
	 *
     */
	 
    protected $defer = false;
	
    /**
     *
	 * @brief Perform post-registration booting of services
     *
	 *
     * @return void
	 *
     */
	 
    public function boot()
    {
		$this->loadViewsFrom(__DIR__.'/Views', 'logs.views');
		
        $this->mergeConfigFrom(__DIR__.'/../../../../config/logs.php', 'logs.config');
    }
	
	/**
	 *  
	 *  @brief Registered routes for the application
	 *  
	 *  @param Router $routers
	 *  
	 *  @return void
	 *  
	 */
	
    public function Routes(Router $routers)
    {		
        $routers->group(['namespace' => 'dev4born\logs\Http\Controllers'], function($routers) {
			
            require __DIR__.'/Http/web.php';
			
        });
    }
	
    /**
     *
	 * @brief Register any package services
     *
	 *
     * @return void
	 *
     */
	 
    public function register()
    {
        $this->Routes($this->app->router);
    }	
}

src/Http/Controllers/LogsActionsController.php
<?php

/**
 *  
 *  Copyright (C) 2017
 *  
 *  File      : LogsActionsController.php
 *  Author    : Miłosz Nowak
 *  Copyright : (c) 2017 Dev4Born
 *  Link      : https://dev4born.pro
 *  Date      : 12/14/17
 *  
 */
 
namespace dev4born\logs\Http\Controllers;

use Illuminate\Routing\Controller;
use Illuminate\Http\Request;

/**
 *  
 *  Class LogsActionsController
 *  
 *  
 *  @package dev4born\logs
 *  
 */
 
class LogsActionsController extends Controller
{
	/**
	 *  
	 *  @brief Create a new controller instance
	 *  
	 *  
	 *  @return void
	 *  
	 */
	
	public function __construct()
	{
		$this->hash = sha1(date('Y-m-d H:i'));
	}
	
	/**
	 *  
	 *  @brief Download log file
	 *  
	 *  @param string $filename
	 *  
	 *  @return Response (download)
	 *  
	 */
	
	public function download(string $filename)
	{
		if($this->token($filename)[0] != $this->hash) {
			
            return back();		
		
		}
		
		$files = storage_path().'/logs/'.$this->token($filename)[1];
		
		return response()->download($files);
	}	
	
	/**
	 *  
	 *  @brief Deleting log file 
	 *  
	 *  @param string $filename
	 *  
	 *  @return Back 
	 *  
	 */
	
	public function remove(string $filename)
	{
		if($this->token($filename)[0] != $this->hash) {
			
            return back();		
		
		}
		
	    $logs = storage_path('logs').'/'.$this->token($filename)[1];	
		
		app('files')->delete($logs);
		
		return back();
	}
	
	/**
	 *  
	 *  @brief Verification of correctness token
	 *  
	 *  @param string $filename
	 *  
	 *  @return $verification
	 *  
	 */
	
	private function token(string $filename) 
	{
		$verification = explode('_', $filename);
		
		return $verification;
	}
}

src/Http/Controllers/LogsController.php
<?php

/**
 *  
 *  Copyright (C) 2017
 *  
 *  File      : LogsController.php
 *  Author    : Miłosz Nowak
 *  Copyright : (c) 2017 Dev4Born
 *  Link      : https://dev4born.pro
 *  Date      : 12/14/17
 *  
 */
 
namespace dev4born\logs\Http\Controllers;

use Illuminate\Support\Facades\Route;
use Illuminate\Routing\Redirector;
use Illuminate\Routing\Controller;
use Illuminate\Http\Request;

/**
 *  
 *  Class LogsController
 *  
 *  
 *  @package dev4born\logs
 *  
 */
 
class LogsController extends Controller
{
	/**
	 *  
	 *  @brief Extension of log files
	 *  
	 *  
	 *  @var string
	 *  
	 */
	
	private $enlargement = 'log';
	
	/**
	 *  
	 *  @brief Create a new controller instance
	 *  
	 *  @param Request $request
	 *  @param Redirector $redirect
	 *  
	 *  @return void
	 *  
	 */
	
	public function __construct(Request $request, Redirector $redirect)
	{
		$this->routers = $request->route('secret');
		
		if($this->permission($this->routers) != true) {
			
			$redirect->to('/')->send();

        } elseif($this->routers == 'view') {
			
			$this->middleware($this->permission($this->routers));
			
		} elseif($this->permission($this->routers) != $this->routers) {
			
		    $redirect->to('/')->send();

        } 
	}
	
	/**
	 *  
	 *  @brief Show all logs/events
	 *  
	 *  @param Request $request
	 *  
	 *  @return View
	 *  
	 */
	
	public function logs(Request $request)
	{
		$list = $this->files();
		
		if($this->files($request['file']) == true && preg_match('/'.$this->enlargement.'/', $request['file'])) {
		
		    $logs = file_get_contents(storage_path().'/logs/'.$request['file']);
		
		} elseif($this->files('laravel.'.$this->enlargement) == true) {
			
		    $logs = file_get_contents(storage_path().'/logs/laravel.'.$this->enlargement);
		
		} else {
			
		    $logs = false;	
			
		}
	
		return view('logs.views::logs', ['request' => $request, 
		                                 'logs'    => $logs,
		                                 'list'    => $list]);	
	
	}

	/**
	 *  
	 *  @brief Verifying that everything functions properly
	 *  
	 *  
	 *  @return Response (JSON)
	 *  
	 */
	
	public function json()
	{
		if($this->files() != true) {
			
			$data = ['error' => 'Everything functions properly.'];
			
		} else {
		
		    $data = ['error' => 'Perhaps an error occurred - check the logs/events.'];
		  
		}
		
		return response()->json($data);
	}
	
	/**
	 *  
	 *  @brief List of files with logs
	 *  
	 *  @parm string $basic
	 *  
	 *  @return Array
	 *  
	 */
	
	private function files($basic = '*')
	{
        $files = glob(storage_path().'/logs/'.$basic);
		
        return array_values($files);		
	}
	
	/**
	 *  
	 *  @brief Permissions/middleware
	 *  
	 *  @parm string $secret
	 *   
	 *  @return Array
	 *  
	 */
	
	private function permission(string $secret)
	{
        $configs = config('logs.config'); 
		 
        $permissions = array();
		
		foreach ($configs as $config) {
			
			array_push($permissions, $config); 

		}
			
		if($permissions[0]['secret'] != $secret) {
			
			$parameter = $permissions[0]['middleware'];
				
		} else {
			
			$parameter = true;
		}
		
		return $parameter;
	}	
}